### PR TITLE
#9527 [S2S] Add a duplicate detection warning when sharing a case wit…

### LINKED ISSUE #9527: [S2S] Add a duplicate detection warning when sharing a case with another instance [5]
### Problem Description

When sharing a case via S2S between one system (source system) and another system (target system) currently there’s no check in place if the case, contact or person that is about to be accepted already exists in the target system.

This could potentially lead to duplicates in the system and multiple entries for the same case being reported to RKI.

For S2S to work properly and sync changes to entities it is crucial for the entities to have the same UUID in both systems. Which means we can’t just simply use the same process for duplicate detection as we do when creating new entities through the UI (the user selects a case or person from a list and uses that one instead). This could potentially lead to the S2S connection breaking because the user can select an entity with a different UUID.

### Proposed Change

When accepting a case that was send via S2S the target system needs to run the normal duplicate detection feature for the entities that are about to be accepted. If at least one potential duplicate has been found, instead of showing a list of potential duplicate candidates the system just shows a warning message, prompting the user to manually check for duplicates and merge them.

In order to prevent double input and double report to RKI, the target system has to check if case data is existing. Therefore, the duplicate detection mechanism have to check the data in target system. The problem is that the place of stay and/or responsible region from the source system is different to the data in target system. Therefore, in first place the target system has to check the person before suggesting an associated case. After indicating a person and associated case in target system it is necessary to inform the user that he is about to accept a case that is already in target system and it is possible to merge cases afterwards.

If case is shared with target system it is still possible to change person data in source and target system. Case data is not affected.

The merge has to be limited. In order to prevent double share after accepting the case in the target system the merge should be only possible for accepted shared case. Otherwise the person uuid and case uuid gets overwrite by non-shared case and it would be possible to revoke share and share case again with target system.

#### Checks need to be executed as follows (Point of view: target system)
- User accepts a share request from a source system
- Target system receives full data set from source system
- Entities are created inside target system
- Run duplicate check for case / contact (depending on what was shared)
- Save number of possible duplicates
- Run duplicate check for person
- Add number of possible duplicates to the previous number
If the sum of possible duplicates found is greater than 0
   Yes: Show warning message with appropriate message (see Add. Information)
else
   No: No action necessary

#### Duplicate warning opens and shows following options:
- OK --> Closes dialog and case gets imported in the target system
- Cancel --> Closes dialog and case gets not imported in target system
- OK and continue to merge overview --> Closes dialog, case gets accept in target system and user gets routed to merging overview in order to merge cases according to just accepted case in target system.

After accepting the case in target system and potential merge of the case the person data can still be updated and gets synchronized in both system. The merged case has also to be marked as shared case (not even in the case overwiew, also in the search overview). There should be a filter setting for shared cases (even if they are merged afterwards). 
If person merge function is implement afterwards the synchronisation afterwards will not work anymore if person is merged in source or target system. This has to be applied and noted for developement of person merging.
 
### Additional Information

Warning messages in german:
- Kein Duplikat erkannt.
- In Ihrem System befindet sich bereits eine Information zur Person. Die Person wird durch das Aufnehmen
nicht in einen Fall konvertiert.
- In Ihrem System befindet sich bereits ein Kontakt zu dieser Person.
- In Ihrem System befindet sich bereits ein Fall zu dieser Person. Sollten Sie die Anfrage annehmen,
ist der Fall als Duplikat in Ihrem System.
Sie sollten nach der Annahme die Fälle konsolidieren. Bitte achten Sie darauf, dass Sie dazu den angenommen Fall
verwenden. Andernfalls kann es zu einer erneute Übermittlung führen.
- In Ihrem System befindet sich bereits ein Kontakt und Fall zu dieser Person.
Sollten Sie den Fall annehmen, ist die Konvertierung des Kontaktes nicht mehr notwendig,
sofern Sie die Konvertierung schon vorgenommen haben.
- In Ihrem System befindet sich bereits eine Person und ein Kontakt mit dem gleichen Namen.
Nachdem Sie den Fall aufgenommen haben, ist es unter Umständen notwendig, eine manuelle
Konvertierung des Kontakts zu dem angenommenen Fall vorzunehmen.
Sie sollten nach der Annahme die Fälle konsolidieren. Bitte achten Sie darauf, dass Sie dazu den angenommen Fall
verwenden. Andernfalls kann es zu einer erneute Übermittlung führen.
Nachdem Sie den Fall aufgenommen haben, ist es unter Umständen notwendig, eine manuelle
Konvertierung des Kontakts zu dem angenommenen Fall vorzunehmen.

Flow diagram scope in target system:
S2S_Process_Accept_Target_System

#### Mockups:
<details>
<summary>Mockups here</summary>

DW_G5
DW_G6
DW_G1
DW_G2
DW_G3
DW_G4

</details>

### PR DESCRIPTION
**Pull request by leventegal-she**
…h another instance

Fixes #9527

### CODE REVIEW

**Review by JonasCir — CHANGES_REQUESTED** [review]

File: sormas-backend/src/main/java/de/symeda/sormas/backend/contact/ContactFacadeEjb.java [hunk 1]
```diff
@@ -1929,17 +1929,57 @@ public List<SimilarContactDto> getMatchingContacts(ContactSimilarityCriteria cri
...
+		}
+
+		return contacts;
+	}
+
+	public boolean hasSimilarContacts(ContactSimilarityCriteria criteria) {
+		return service.exists((cb, root, cq) -> getSimilarityFilters(criteria, cb, root, new ContactQueryContext(cb, cq, root)));
+	}
+
+	private Predicate getSimilarityFilters(
```
[the remark this answers is not in the document]
Comment (leventegal-she) [reply sormas-backend/src/main/java/de/symeda/sormas/backend/contact/ContactFacadeEjb.java:1955]: I don't know what would be a the best place for this, this is just a part of the `getMatchingContacts` method, extracted into e separate method to be reusable. For cases the corresponding method was in the service that's why `getSimilarityFilters` was created there

File: sormas-ui/src/main/java/de/symeda/sormas/ui/sormastosormas/SormasToSormasController.java [hunk 2]
```diff
@@ -203,12 +204,99 @@ public void acceptShareRequest(ShareRequestIndexDto request, Runnable callback)
...
+						Strings.messageSormasToSormasSimilarConvertedCaseFound,
+						Captions.actionOkAndGoToMerge,
+						() -> ControllerProvider.getCaseController().navigateToMergeCasesView());
+					break;
+				case CONTACT_TO_CASE:
+					confirmDuplicateFound(
+						request,
+						Strings.messageSormasToSormasSimilarContactToCaseFound,
+						Captions.actionOkAndGoToContactDirectory,
+						// TODO - filter
```
[the remark this answers is not in the document]
Comment (leventegal-she) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/sormastosormas/SormasToSormasController.java]: removed todo, and created a ticket https://github.com/hzi-braunschweig/SORMAS-Project/issues/10569 as suggested in https://github.com/hzi-braunschweig/SORMAS-Project/issues/9527#issuecomment-1268008512

File: sormas-backend/src/main/java/de/symeda/sormas/backend/contact/ContactFacadeEjb.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (JonasCir) [reply sormas-backend/src/main/java/de/symeda/sormas/backend/contact/ContactFacadeEjb.java:1955]: fine with me

File: sormas-ui/src/main/java/de/symeda/sormas/ui/sormastosormas/SormasToSormasController.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (JonasCir) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/sormastosormas/SormasToSormasController.java]: thx

**Review by JonasCir — APPROVED** [review]
lgtm

**Pull request merged**